<a href="https://colab.research.google.com/github/izzat-ai/learning-ai/blob/main/numpy/climate_change.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

In [2]:
temperatures = np.array([
    15, 17, 22, 28, 33, 38, 40, 39, 34, 27, 20, 16, # Shahar A
    10, 12, 18, 24, 30, 35, 37, 36, 31, 23, 15, 11, # Shahar B
    25, 26, 28, 30, 32, 33, 34, 34, 33, 31, 28, 26  # Shahar C (Tropik)
])
temperatures.size

36

- View each city's data in a separate row

In [3]:
data = temperatures.reshape(3, 12)
data

array([[15, 17, 22, 28, 33, 38, 40, 39, 34, 27, 20, 16],
       [10, 12, 18, 24, 30, 35, 37, 36, 31, 23, 15, 11],
       [25, 26, 28, 30, 32, 33, 34, 34, 33, 31, 28, 26]])

- Temperatures above 35 degrees Celsius are considered "Abnormally hot" due to global warming
- Filter only temperatures above 35 degrees Celsius in City B (City 2)

In [4]:
data[data>35]

array([38, 40, 39, 37, 36])

In [5]:
city2_anomal = data[1][data[1]>35]
city2_anomal

array([37, 36])

- When analyzing the data for City B, it was found that there were a total of 2 cases where the temperature exceeded 35 degrees Celsius. This is the hottest time of the summer for that area

In [6]:
# convert temperature from Celsius to Fahrenheit
temps_f = data*1.8 + 32
temps_f

array([[ 59. ,  62.6,  71.6,  82.4,  91.4, 100.4, 104. , 102.2,  93.2,
         80.6,  68. ,  60.8],
       [ 50. ,  53.6,  64.4,  75.2,  86. ,  95. ,  98.6,  96.8,  87.8,
         73.4,  59. ,  51.8],
       [ 77. ,  78.8,  82.4,  86. ,  89.6,  91.4,  93.2,  93.2,  91.4,
         87.8,  82.4,  78.8]])

In [7]:
# determine the average summer temperature for each city
np.mean(temps_f[:, [5, 6, 7]], axis=1)

array([102.2,  96.8,  92.6])

In [8]:
# transpose each city to compare their temperatures
fahren = temps_f.T
fahren

array([[ 59. ,  50. ,  77. ],
       [ 62.6,  53.6,  78.8],
       [ 71.6,  64.4,  82.4],
       [ 82.4,  75.2,  86. ],
       [ 91.4,  86. ,  89.6],
       [100.4,  95. ,  91.4],
       [104. ,  98.6,  93.2],
       [102.2,  96.8,  93.2],
       [ 93.2,  87.8,  91.4],
       [ 80.6,  73.4,  87.8],
       [ 68. ,  59. ,  82.4],
       [ 60.8,  51.8,  78.8]])

In [9]:
# determining the hottest city in January
np.argmax(fahren[0, :])

np.int64(2)

In [14]:
# The temperature in all cities will increase by 2% next year. Predicting data for next year
new_year_temp = temps_f*1.02
new_year_temp

array([[ 60.18 ,  63.852,  73.032,  84.048,  93.228, 102.408, 106.08 ,
        104.244,  95.064,  82.212,  69.36 ,  62.016],
       [ 51.   ,  54.672,  65.688,  76.704,  87.72 ,  96.9  , 100.572,
         98.736,  89.556,  74.868,  60.18 ,  52.836],
       [ 78.54 ,  80.376,  84.048,  87.72 ,  91.392,  93.228,  95.064,
         95.064,  93.228,  89.556,  84.048,  80.376]])

In [15]:
# Combining old information and new predictions into one
data_3d = np.dstack((data, new_year_temp))
data_3d

array([[[ 15.   ,  60.18 ],
        [ 17.   ,  63.852],
        [ 22.   ,  73.032],
        [ 28.   ,  84.048],
        [ 33.   ,  93.228],
        [ 38.   , 102.408],
        [ 40.   , 106.08 ],
        [ 39.   , 104.244],
        [ 34.   ,  95.064],
        [ 27.   ,  82.212],
        [ 20.   ,  69.36 ],
        [ 16.   ,  62.016]],

       [[ 10.   ,  51.   ],
        [ 12.   ,  54.672],
        [ 18.   ,  65.688],
        [ 24.   ,  76.704],
        [ 30.   ,  87.72 ],
        [ 35.   ,  96.9  ],
        [ 37.   , 100.572],
        [ 36.   ,  98.736],
        [ 31.   ,  89.556],
        [ 23.   ,  74.868],
        [ 15.   ,  60.18 ],
        [ 11.   ,  52.836]],

       [[ 25.   ,  78.54 ],
        [ 26.   ,  80.376],
        [ 28.   ,  84.048],
        [ 30.   ,  87.72 ],
        [ 32.   ,  91.392],
        [ 33.   ,  93.228],
        [ 34.   ,  95.064],
        [ 34.   ,  95.064],
        [ 33.   ,  93.228],
        [ 31.   ,  89.556],
        [ 28.   ,  84.048],
        [ 26.   

- City B has a very high temperature volatility of 10 to 37. This can be a challenge for agriculture .
- The temperature variation in City C is very small (between 25 and 34 degrees Celsius). It is the most suitable city for tourism .
- Next year's forecast predicts that the hottest temperature in City A will exceed 106 Fahrenheit. This should be a warning to the health system .